File: intellij-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/ResolveCommand.kt
package com.ezevillo.gitreview.domain

data class ResolvedCommand(
    val command: String,
    val args: List<String>,
)

private val WINDOWS_NATIVE_EXECUTABLE = Regex("""\.(exe|cmd|bat)$""", RegexOption.IGNORE_CASE)

/**
 * Resolves executable + args for a git-review invocation
 * (contracts/cli-invocation.md § Forma de toda invocación).
 *
 * @param platform `"win32"` or other (POSIX).
 */
fun resolveCommand(
    verb: String,
    args: List<String>,
    gitReviewPath: String?,
    platform: String = System.getProperty("os.name").lowercase().let {
        if (it.contains("win")) "win32" else "posix"
    },
): ResolvedCommand {
    if (gitReviewPath.isNullOrBlank()) {
        return ResolvedCommand(command = "git", args = listOf("review", verb) + args)
    }
    if (platform == "win32" && !WINDOWS_NATIVE_EXECUTABLE.containsMatchIn(gitReviewPath)) {
        return ResolvedCommand(command = "sh", args = listOf(gitReviewPath, verb) + args)
    }
    return ResolvedCommand(command = gitReviewPath, args = listOf(verb) + args)
}

File: visualstudio-extension/src/GitReview.Domain/ResolveCommand.cs
using System.Text.RegularExpressions;

namespace GitReview.Domain;

public sealed record ResolvedCommand(string Command, IReadOnlyList<string> Args);

public static class ResolveCommand
{
    private static readonly Regex WindowsNativeExecutable =
        new(@"\.(exe|cmd|bat)$", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>
    /// Resolves executable + args for a git-review invocation
    /// (contracts/cli-invocation.md).
    /// </summary>
    /// <param name="platform">"win32" or other (POSIX).</param>
    public static ResolvedCommand Resolve(
        string verb,
        IReadOnlyList<string> args,
        string? gitReviewPath,
        string? platform = null)
    {
        platform ??= RuntimeInfo.IsWindows ? "win32" : "posix";
        if (string.IsNullOrWhiteSpace(gitReviewPath))
            return new ResolvedCommand("git", new[] { "review", verb }.Concat(args).ToList());
        if (platform == "win32" && !WindowsNativeExecutable.IsMatch(gitReviewPath))
            return new ResolvedCommand("sh", new[] { gitReviewPath, verb }.Concat(args).ToList());
        return new ResolvedCommand(gitReviewPath, new[] { verb }.Concat(args).ToList());
    }
}
